src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -0,0 +1,517 @@
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    ChipLogProgress(Zcl, "DEM: %s Good PowerAdjustRequest() args.", __FUNCTION__);
+
+    /* Call the delegate to do the power adjustment */
+    status = mDelegate.PowerAdjustRequest(power, durationSec);
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    if (status != Status::Success)
+    {
+        ChipLogError(Zcl, "DEM: %s Failed to PowerAdjustRequest() args.", __FUNCTION__);
+    }
+    return;
+}
+
+void Instance::HandleCancelPowerAdjustRequest(HandlerContext & ctx,
+                                              const Commands::CancelPowerAdjustRequest::DecodableType & commandData)
+{
+    Status status = Status::Success;
+    ESAStateEnum esaStatus;
+
+    /* Check that the ESA state is PowerAdjustActive */
+    esaStatus = mDelegate.GetESAState();
+    if (ESAStateEnum::kPowerAdjustActive != esaStatus)
+    {
+        ChipLogError(Zcl, "DEM: %s - kPowerAdjustActive != esaStatus", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* Call the delegate to cancel the power adjustment */
+    status = mDelegate.CancelPowerAdjustRequest();
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    if (status != Status::Success)
+    {
+        ChipLogError(Zcl, "DEM: %s Failed to CancelPowerAdjustRequest()", __FUNCTION__);
+        return;
+    }
+
+    return;
+}
+
+void Instance::HandleStartTimeAdjustRequest(HandlerContext & ctx,
+                                            const Commands::StartTimeAdjustRequest::DecodableType & commandData)
+{
+    Status status                                               = Status::Success;
+    uint32_t earliestStartTime                                  = 0;
+    uint32_t latestEndTime                                      = 0;
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
+
+    uint32_t duration;
+    uint32_t requestedStartTime = commandData.requestedStartTime;
+
+    if (forecast.IsNull())
+    {
+        ChipLogError(Zcl, "DEM: %s - Forecast is Null", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    if (ESAStateEnum::kUserOptOut == mDelegate.GetESAState())
+    {
+        ChipLogError(Zcl, "DEM: %s - ESAState = kUserOptOut", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* If the RequestedStartTime value resulted in a time shift which is
+     * outside the time constraints of EarliestStartTime and
+     * LatestEndTime, then the command SHALL be rejected with CONSTRAINT_ERROR;
+     * otherwise the command SHALL be rejected with FAILURE
+     */
+    /* earliestStartTime is optional based on the FA (ForecastAdjust) feature AND can be nullable */
```

Comment:
... AND can be nullable
=>   is nullable

Reply:
changed.